docs: add new blog(web scraping expert) (#2731)
apify/crawlee: website/blog/2024/11-10-web-scraping-tips/index.md

So, let's explore the mindset of a web scraping developer.

![How to think like a web scraping expert](./img/scraping-tips.webp)

<!-- truncate -->

## 1. Choosing a data source for the project

When you start working on a project, you likely have a target site from which you need to extract specific data. Check what possibilities this site or application provides for data extraction. Here are some possible options:

- `Official API` - the site may provide a free official API through which you can get all the necessary data. This is the best option for you. For example, you can consider this approach if you need to extract data from [`Yelp`](https://docs.developer.yelp.com/docs/fusion-intro)
- `Website` - in this case, we study the website, its structure, as well as the ways the frontend and backend interact
- `Mobile Application` - in some cases, there's no website or API at all, or the mobile application provides more data, in which case, don't forget about the [`man-in-the-middle`](https://blog.apify.com/using-a-man-in-the-middle-proxy-to-scrape-data-from-a-mobile-app-api-e954915f979d/) approach

If one data source fails, try accessing another available source.

> For example, for `Yelp`, all three options are available, and if the `Official AP`I doesn't suit you for some reason, you can try the other two.

Reviewer: ```suggestion
For example, for `Yelp`, all three options are available, and if the `Official API` doesn't suit you for some reason, you can try the other two.
```

- `robots` is the established name for crawlers in SEO. Usually, this refers to crawlers of major search engines like Google and Bing, or services like Ahrefs and ChatGPT.
- `robots.txt` is a file describing the allowed behavior for robots. It includes permitted crawler user-agents, wait time between page scans, patterns of pages forbidden for scanning, and more. These rules are typically based on which pages should be indexed by search engines and which should not.
- `sitemap` describes the site structure to make it easier for robots to navigate. It also helps in scanning only the content that needs updating, without creating unnecessary load on the site

Since you're not [`Google`](http://google.com/) or any other popular search engine, the robot rules in `robots.txt` will likely be against you. But combined with the `sitemap`, this is a good place to study the site structure, expected interaction with robots, and non-browser user-agents. In some situations, it simplifies data extraction from the site.

> For example, using the [`sitemap`](https://www.coolbrnoblog.cz/wp-sitemap.xml) for [the blog](http://www.coolbrnoblog.cz), you can easily get direct links to posts both for the entire lifespan of the blog and for a specific period. One simple check, and you don't need to implement pagination logic.

Reviewer: I'd rather link to some servers we can control (so the links cannot break eventually). What about https://crawlee.dev and https://crawlee.dev/sitemap.xml ? :)

- You need to pass some parameters, for example cookies, or specific technical headers
- The site requires that when accessing this endpoint, there is a corresponding `Referrer` header
- The site expects that the headers will follow a certain order. I've encountered this only a couple of times, but I have encountered it
- The site uses protection against web scraping, for example with `TLS fingerprint`

And many other possible reasons, each of which requires separate analysis.

## 10. Experiment with request parameters

Explore what results you get when changing request parameters, if any. Some parameters may be missing but supported on the server side. For example, `order`, `sort`, `per_page`, `limit`, and others. Try adding them and see if the behavior changes.

This is especially relevant for sites using [`graphql`](https://graphql.org/)

Let's consider this [`example`](https://restoran.ua/en/posts?subsection=0)

If you analyze the site, you'll see a request that can be reproduced with the following code, I've formatted it a bit to improve readability:

```python
import requests

> url = "<https://restoran.ua/graphql>"

Reviewer: This snippet fails with `Invalid schema`, this doesn't seem right:
```suggestion
url = "https://restoran.ua/graphql"
```

data = {
    "operationName": "Posts_PostsForView",
    "variables": {"sort": {"sortBy": ["startAt_DESC"]}},
    "query": """query Posts_PostsForView(
    $where: PostForViewWhereInput,
    $sort: PostForViewSortInput,
    $pagination: PaginationInput,
    $search: String,
    $token: String,

>     $coordinates_slice: SliceInput,

Reviewer: this snippet throws a GraphQL syntax error, likely because of this
```suggestion
    $coordinates_slice: SliceInput)
```

    {
        PostsForView(
                where: $where
                sort: $sort
                pagination: $pagination
                search: $search
                token: $token
                ) {
                        id
                        title: ukTitle
                        summary: ukSummary
                        slug
                        startAt
                        endAt
                        newsFeed
                        events
                        journal
                        toProfessionals
                        photoHeader {
                            address: mobile
                            __typename
                            }
                        coordinates(slice: $coordinates_slice) {
                            lng
                            lat
                            __typename
                            }
                        __typename
                    }
    }"""
}

response = requests.post(url, json=data)

print(response.json())
```

Now I'll update it to get results in 2 languages at once, and most importantly, along with the internal text of the publications:

> ```python

Reviewer: I'd rather see what was added than see the whole script again... consider the `diff` Markdown syntax highlighter or Docusaurus' [Line Highlighting](https://docusaurus.io/docs/markdown-features/code-blocks#line-highlighting) (I'd actually prefer this).

import requests

> url = "<https://restoran.ua/graphql>"

Reviewer: Same here

```suggestion
url = "https://restoran.ua/graphql"
```

}

response = requests.post(url, json=data)

> # highlight-start

Reviewer: I think this is more of a debugging relict (it would make more sense to call `.json()` here too - the response is still JSON document). Feel free to swap it to `.json()` (and remove the highlighting).
